### ISSUE TITLE
`Out` type of `ProductLength` is not a `Nat`

### ISSUE THREAD

**Issue by AndreasKostler** [milessabin/shapeless#624]
This won't compile

``` scala
trait WantsNat[N <: Nat]

def foo[P <: Product](p: P)(implicit pl: ProductLength[P]): WantsNat[pl.Out] = new WantsNat[pl.Out] {}
```

**Comment by milessabin** [comment]
Good catch. There is a workaround though,
```scala
def foo[P <: Product](p: P)(implicit pl: ProductLength[P] { type Out <: Nat }): WantsNat[pl.Out] =
  new WantsNat[pl.Out] {}
```

**Comment by AndreasKostler** [comment]
That works, kinda breaks symmetry with `HList` `Length` though.

**Comment by milessabin** [comment]
Agreed ... it's a workaround, not the right answer :-)

**Comment by milessabin** [comment]
The fix breaks binary compatibility so this is going to have to wait for 2.4.0.

**Comment by Ssstlis** [comment]
Should we close this issue 'cause of merged #973 ?

**Comment by joroKr21** [comment]
Oh yes thanks fixed by #973 👍
